### DISCUSSION TITLE
Occupancy grid for Ray Sampling in NeRFs

### DISCUSSION THREAD

**Discussion by peerschuett** [NVlabs/instant-ngp#377] [Q&A] [answered]
Hi, 
thanks for you amazing work! I am trying to use your occupancy grid, that is explained in Section C.2 of your paper, for one of my models. But I am not sure how it exactly works. I understand the basic idea, that you want to skip empty cells and only consider occupied ones, but during the first training iterations the NeRF doesn't know what areas are occupied and what not, because the density values output from it are too small. If you would therefore update the grid immediately with the density values of this NeRF you wouldn't get any samples anymore, because the NeRF thinks that nearly everything is unoccupied (if I am not mistaken). How do you therefore decide when the network has enough knowledge?

**Comment by neilgogogo** [comment]
Hi, in the very beginning, the author firstly uses the mark_untrained_density_grid func to filter out the unseen grids, marking the unseen grid as -1 and seen grid 0. 

And when the iteration is 0 at the very beginning, the mlp_output is 0, but after exponential activation, the mlp is 1. Then the grid_out, which is density_grid_tmp is scalbnf(MIN_CONE_STEPSIZE(), level) (I cannot understand the __float_as_uint operation in the code).

<img width="1281" alt="image" src="https://user-images.githubusercontent.com/53148705/159147357-157f67f6-e6aa-4750-8770-7d16db3652f4.png">


Then in the ema_grid_samples_nerf function, the density_grid_tmp value will be assign to density_grid=0 part, and density_grid=-1 part will be unchanged. 

<img width="814" alt="image" src="https://user-images.githubusercontent.com/53148705/159147231-d6f514a1-b05c-448b-b601-cd6aa534c4e7.png">

So perhaps because of the exponential activation, the mlp is 1 although the mlp_out=0.

**Comment by Tom94** [comment, marked as answer]
Hi there, there exists a safeguard against zero-everywhere occupancy here: https://github.com/NVlabs/instant-ngp/blob/8625c253ed26f10d8a071f928da48ecad08b472d/src/testbed_nerf.cu#L562

The threshold for setting a cell as occupied may never be larger than the mean density.
